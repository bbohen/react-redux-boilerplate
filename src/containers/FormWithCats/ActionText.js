import styled from 'styled-components';

const ActionText = styled.div`
  color: ${props => props.theme.colors.secondary};
  cursor: pointer;
`;

export default ActionText;
