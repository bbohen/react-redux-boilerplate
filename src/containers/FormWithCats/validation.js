/**
 * validation logic for FormWithCats
 * if these need to be shared a central validation file should be created
 */

const validation = {
  email: value => (value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value) ? 'Thats not an email!' : undefined),
  number: value => (value && isNaN(Number(value))) ? 'Thats not a number!' : undefined, // eslint-disable-line
  required: value => (value ? undefined : 'Required!'),
};

export default validation;
