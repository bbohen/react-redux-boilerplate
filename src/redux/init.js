import { compose, createStore, applyMiddleware } from 'redux';
import { middleware as reduxPackMiddleware } from 'redux-pack';
import thunk from 'redux-thunk';

import rootReducer from './reducer';

/* eslint-disable no-underscore-dangle  */
const composeEnhancers = process.env.NODE_ENV !== 'production' &&
(window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose);
/* eslint-enable */

export default () => createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(thunk, reduxPackMiddleware)),
);
