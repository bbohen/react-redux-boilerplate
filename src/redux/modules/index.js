export cats from './cats';
export loremIpsum from './loremIpsum';
