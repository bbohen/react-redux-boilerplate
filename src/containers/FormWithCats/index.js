import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm } from 'redux-form';

import { FormField } from 'components';
import { clearRandomCat, giveRandomCat } from '../../redux/modules/cats';
import validation from './validation';
import ActionText from './ActionText';
import CatFormHeading from './Heading';
import CatFormButton from './Button';
import Form from './Form';
import Wrapper from './Wrapper';

const UnconnectedFormWithCats = (props) => {
  const { email, number, required } = validation;
  let result;

  if (props.randomCat) {
    result = (
      <div>
        <CatFormHeading>
          Here is your cat.
          <ActionText onClick={props.clearRandomCat}>
            Try again?
          </ActionText>
        </CatFormHeading>
        <img
          alt=""
          src={props.randomCat}
        />
      </div>);
  }

  return (
    <Wrapper>
      {!result &&
      <Form onSubmit={props.handleSubmit && props.handleSubmit(props.giveRandomCat)}>
        <CatFormHeading>Form.</CatFormHeading>
        <Field
          label="Name:"
          name="name"
          component={FormField}
          type="text"
          validate={required}
        />
        <Field
          label="Email:"
          name="email"
          component={FormField}
          type="text"
          validate={[email, required]}
        />
        <Field
          label="Number:"
          name="number"
          component={FormField}
          type="text"
          validate={[number, required]}
        />
        <CatFormButton type="submit">
            Submit the form.
        </CatFormButton>
      </Form> }
      {result}
    </Wrapper>
  );
};

UnconnectedFormWithCats.propTypes = {
  clearRandomCat: PropTypes.func.isRequired,
  giveRandomCat: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  randomCat: PropTypes.string,
};

UnconnectedFormWithCats.defaultProps = {
  randomCat: '',
};

export default connect(
  ({ cats = {} }) => ({
    randomCat: cats.randomCat,
  }),
  {
    clearRandomCat,
    giveRandomCat,
  },
)(reduxForm({
  form: 'cats',
})(UnconnectedFormWithCats));
