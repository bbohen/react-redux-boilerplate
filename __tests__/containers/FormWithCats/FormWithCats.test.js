import React from 'react';
import { reducer as formReducer } from 'redux-form';
import { createStore, combineReducers } from 'redux';
import { Provider } from 'react-redux';

import FormWithCats from '../../../src/containers/FormWithCats';
import cats from '../../../src/redux/modules/cats';
import validation from '../../../src/containers/FormWithCats/validation';
import testRendererWithProvider from '../../testUtils/testRendererWithProvider';

describe('<FormWithCats />', () => {
  let store;
  let WrappedFormWithCats;
  let component;
  let tree;

  it('renders the FormWithCats container integrated with redux form', () => {
    store = createStore(combineReducers({ form: formReducer }));
    WrappedFormWithCats = (
      <Provider store={store}>
        <FormWithCats />
      </Provider>
    );
    component = testRendererWithProvider(WrappedFormWithCats);
    tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders the FormWithCats container with a random cat provided by redux', () => {
    store = createStore(
      combineReducers({ cats, form: formReducer }),
      {
        cats: {
          randomCat: 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif',
        },
      },
    );
    WrappedFormWithCats = (
      <Provider store={store}>
        <FormWithCats />
      </Provider>
    );
    component = testRendererWithProvider(WrappedFormWithCats);
    tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('<FormWithCats /> validation methods', () => {
  const { email, number, required } = validation;

  describe('email validation', () => {
    it('allows an email address', () => {
      expect(email('test@test.com')).toBeUndefined();
      expect(email('test@test.edu')).toBeUndefined();
      expect(email('test@test.org')).toBeUndefined();
      expect(email('test@test.gov')).toBeUndefined();
    });

    it('requires an email address', () => {
      const validationMessage = 'Thats not an email!';
      expect(email('asd')).toEqual(validationMessage);
      expect(email('123')).toEqual(validationMessage);
      expect(email('asd123@')).toEqual(validationMessage);
      expect(email('asd123@test')).toEqual(validationMessage);
      expect(email('asd123@test.')).toEqual(validationMessage);
      expect(email('asd123@test.c')).toEqual(validationMessage);
    });
  });

  describe('number validation', () => {
    it('allows a number', () => {
      expect(number(1)).toBeUndefined();
      expect(number(21)).toBeUndefined();
      expect(number(123)).toBeUndefined();
      expect(number(123131321313)).toBeUndefined();
      expect(number('1')).toBeUndefined();
      expect(number('123')).toBeUndefined();
    });

    it('requires a number', () => {
      const validationMessage = 'Thats not a number!';
      expect(number('test')).toEqual(validationMessage);
      expect(number('1t2')).toEqual(validationMessage);
    });
  });

  describe('required validation', () => {
    const validationMessage = 'Required!';
    it('requires a value', () => {
      expect(required('test')).toBeUndefined();
      expect(required(123)).toBeUndefined();
      expect(required('')).toEqual(validationMessage);
      expect(required(undefined)).toEqual(validationMessage);
      expect(required(null)).toEqual(validationMessage);
    });
  });
});
