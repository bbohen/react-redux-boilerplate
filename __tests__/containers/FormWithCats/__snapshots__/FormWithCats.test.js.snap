// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormWithCats /> renders the FormWithCats container integrated with redux form 1`] = `
<div
  className="eMVHAg"
>
  <form
    className="dZhWbL"
    onSubmit={[Function]}
  >
    <h2
      className="dOqEkG"
    >
      Form.
    </h2>
    <div
      className="bzWcLi"
    >
      <div
        className="bjhZVI"
      >
        Name:
      </div>
      <input
        className="iZQdaK"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder=""
        type="text"
        value=""
      />
    </div>
    <div
      className="bzWcLi"
    >
      <div
        className="bjhZVI"
      >
        Email:
      </div>
      <input
        className="iZQdaK"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder=""
        type="text"
        value=""
      />
    </div>
    <div
      className="bzWcLi"
    >
      <div
        className="bjhZVI"
      >
        Number:
      </div>
      <input
        className="iZQdaK"
        name="number"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder=""
        type="text"
        value=""
      />
    </div>
    <button
      className="YjMsv"
      type="submit"
    >
      Submit the form.
    </button>
  </form>
</div>
`;

exports[`<FormWithCats /> renders the FormWithCats container with a random cat provided by redux 1`] = `
<div
  className="eMVHAg"
>
  <div>
    <h2
      className="dOqEkG"
    >
      Here is your cat.
      <div
        className="fpWVTn"
        onClick={[Function]}
      >
        Try again?
      </div>
    </h2>
    <img
      alt=""
      src="https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif"
    />
  </div>
</div>
`;
